Adding return type to signature causes `Don't use interop for EnsoObject` error

Adding `-> Integer` to the checked return type of `Decimal.to_integer` results in the error below (when running `Decimal_Spec`).

If the `Warning.attach` is removed, the problem goes away, so I imagine that the error we want to get here is something about `Any` is not `Integer`.

(I also tried keeping `Warning.attach` but casting it to Integer, and also got the error below in that case.)

Change on https://github.com/enso-org/enso/tree/bug/gmt/10613.

```
    - [PENDING] should compare correctly to Integer and Float
        Reason: https://github.com/enso-org/enso/issues/10163
Execution finished with an error: java.lang.AssertionError: Don't use interop for EnsoObject: org.enso.interpreter.runtime.error.WithWarnings
        at <java> org.enso.runtime/org.enso.interpreter.runtime.library.dispatch.TypeOfNode$WithoutType$Interop.resolve(TypeOfNode.java:213)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.library.dispatch.TypeOfNode.withoutType(TypeOfNode.java:83)
        at <java> org.enso.runtime/org.enso.interpreter.runtime.library.dispatch.TypeOfNodeGen.execute(TypeOfNodeGen.java:186)
        at <java> org.enso.runtime/org.enso.interpreter.node.expression.builtin.meta.IsValueOfTypeNode.typeAndCheck(IsValueOfTypeNode.java:55)
        at <java> org.enso.runtime/org.enso.interpreter.node.expression.builtin.meta.IsValueOfTypeNode$Untyped.doOther(IsValueOfTypeNode.java:180)
```

--- a/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso
+++ b/distribution/lib/Standard/Base/0.0.0-dev/src/Data/Decimal.enso
@@ -764,8 +764,7 @@ type Decimal
              d = Decimal.new "2345.6"
              d.to_integer
              # => 2345
-    to_integer : Integer
-    to_integer self =
+    to_integer self -> Integer =
         as_biginteger = self.big_decimal.toBigInteger
         back_to_decimal = BigDecimal.new as_biginteger
         are_equal = (self.big_decimal.compareTo back_to_decimal) == 0

--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/library/dispatch/TypeOfNode.java
@@ -74,7 +74,17 @@ public abstract class TypeOfNode extends Node {
     return execute(value.getValue());
   }
 
-  @Specialization(guards = {"!types.hasType(value)"})
+  static boolean isWithoutType(Object value, TypesLibrary types) {
+    if (value instanceof EnsoObject) {
+      return false;
+    }
+    if (types.hasType(value)) {
+      return false;
+    }
+    return true;
+  }
+
+  @Specialization(guards = {"isWithoutType(value, types)"})
   Object withoutType(
       Object value,
       @Shared("interop") @CachedLibrary(limit = "3") InteropLibrary interop,

